dds/DCPS/AbstractSample.h

```diff
@@ -0,0 +1,262 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#ifndef OPENDDS_DCPS_ABSTRACT_SAMPLE_H
+#define OPENDDS_DCPS_ABSTRACT_SAMPLE_H
+
+#include "Serializer.h"
+#include "TypeSupportImpl.h"
+#include "RcHandle_T.h"
+#include "FilterEvaluator.h"
+#include "XTypes/DynamicDataAdapter.h"
+
+#include <dds/DdsDynamicDataC.h>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+class AbstractSample;
+typedef RcHandle<AbstractSample> AbstractSample_rch;
+
+/**
+ * Represents a sample that can either be an instance of a C++ generated type
+ * from opendds_idl or a DynamicData. This is meant to be used by
+ * DataReaderImpl and DataWriterImpl.
+ */
+class OpenDDS_Dcps_Export AbstractSample : public virtual RcObject {
+public:
+  AbstractSample(bool owns_data, bool read_only, bool key_only)
+  : owns_data_(owns_data)
+  , read_only_(read_only)
+  , key_only_(key_only)
+  {
+  }
+
+  virtual ~AbstractSample() {}
+
+  bool owns_data() const
+  {
+    return owns_data_;
+  }
+
+  bool read_only() const
+  {
+    return owns_data_;
+  }
+
+  bool key_only() const
+  {
+    return key_only_;
+  }
+
+  virtual bool serialize(Serializer& ser) const = 0;
+  virtual bool deserialize(Serializer& ser) const = 0;
+  virtual size_t serialized_size(const Encoding& enc) const = 0;
+  virtual bool compare(const AbstractSample& other) const = 0;
+  virtual bool to_message_block(ACE_Message_Block& mb) = 0;
+  virtual bool from_message_block(const ACE_Message_Block& mb) = 0;
+  virtual AbstractSample_rch copy(bool read_only, bool key_only) const = 0;
+  AbstractSample_rch copy(bool read_only) const
+  {
+    return copy(read_only, key_only_);
+  }
+#ifndef OPENDDS_SAFETY_PROFILE
+  virtual DDS::DynamicData* get_dynamic_data(DDS::DynamicType_ptr type) = 0;
```

Comment:
For IDL interface types, let's consistently use either `*` or `_ptr`.... or use `_var`

Reply:
Making it a var would solve your comment on the reference counting and let me undo the changes to make it possible to have a "null" `DyanmicDataAdaptor`.